Use `unix.Ioctl*` on BSD systems, deduplicate `tscreen_*.go`

niten94 opened the pull request:
The behavior that occurs when calling `syscall.Syscall` and passing `syscall.SYS_IOCTL` in OpenBSD has been changed in Go 1.23 because `syscall(2)` is removed in OpenBSD 7.5. There may be changes in any platform where using `syscall.Syscall*` will not work, so `Syscall6` is replaced with `golang.org/x/sys/unix.Ioctl*` in this pull request to partly address zyedidia/micro#3557.

There may be fields in `Termios` added that have to be copied, so the whole struct is also copied before modifying it in this pull request. The serial TTY session in my OpenBSD VM always crashed when I compiled with the original `tscreen_linux.go` code.

I do not know how micro was able to start up properly on old OpenBSD versions even though `unsafe.Pointer` was not converted to `uintptr` in arguments like written in the 4th pattern written in [documentation](https://pkg.go.dev/unsafe@go1.23.4#Pointer), but the screen would be blank in newer versions.

There is a different module that had to be updated but micro seemed to start up properly when I tested using an OpenBSD 7.6 VM. I do not have a MacOS machine so I cannot test if there are no bugs on MacOS.

niten94 committed:
Copy termios struct as whole

JoeKar commented:
Upstream `SYS_IOCTL` isn't used, but still `syscall` is invoked at some places.
The long term target is to keep track with the upstream version of `tcell` and only merge what is necessary for `micro` which isn't upstream so far.

Edit1:
Unfortunately I can't do any merges here, since it is a other private repository from @zyedidia.

As a short term plan it is maybe better to prepare https://github.com/micro-editor/tcell [^see] with the same state as https://github.com/zyedidia/tcell has to have ready2go drop-in replacement for `micro`, with better/extended community access.

Edit2:
See https://github.com/zyedidia/micro/pull/3593 resp. https://github.com/micro-editor/tcell/tree/legacy branch.

[^see]: https://github.com/micro-editor/tcell/tree/upstream-rebase already created on a much newer upstream version with some of `micro`'s patches, but by far not all.

niten94 committed:
tscreen_*.go: Use unix.Ioctl* on BSD systems, deduplicate code

Replace Syscall with unix.Ioctl* in tscreen_{bsd,darwin}.go. Merge
tscreen_{bsd,solaris}.go with tscreen_linux.go and rename to
tscreen_unix.go to deduplicate code.

niten94 commented:
> Upstream `SYS_IOCTL` isn't used, but still `syscall` is invoked at some places.

`SetNonblock` seems to be the only function in `syscall` invoked upstream on Unix platforms, but libc functions seem to be used internally so upstream code may currently work in platforms where syscalls cannot be used. I may still create a pull request upstream to switch to `golang.org/x/sys/unix`.

> The long term target is to keep track with the upstream version of `tcell` and only merge what is necessary for `micro` which isn't upstream so far.

I think it is good to avoid adding code different with upstream, but the code in `tscreen_*.go` is currently moved to `tty_unix*.go` and there will be more changes needed if added. This pull request is made to fix bugs only occuring in the fork without changing APIs and features, so I do not see problems with changing `tscreen_*.go`. However, I realized there is this old version of `tscreen_unix.go` similar with this pull request so I could adjust to it if it is better: https://github.com/gdamore/tcell/blob/e9095fe4f117459e3d961ea4eb3f55bf8f45e91c/tscreen_unix.go

> Edit2: See [zyedidia/micro#3593](https://github.com/zyedidia/micro/pull/3593) resp. https://github.com/micro-editor/tcell/tree/legacy branch.

I have no experience with maintaining forks but I will leave a comment when there is anything I could find or help with deciding.

JoeKar commented:
> I have no experience with maintaining forks but I will leave a comment when there is anything I could find or help with deciding.

We now have two forks[^1][^2]. For us it is easier to get the things into [^2] and [^2] used as dependency for `micro` itself, since we've no rights to merge in [^1].
What I was trying to tell is, that this PR is maybe better to directly target the [legacy](https://github.com/micro-editor/tcell/tree/legacy) of [^2] instead of [^1].

@dmaluka:
What do you think?

[^1]: https://github.com/zyedidia/tcell
[^2]: https://github.com/micro-editor/tcell, where the [legacy](https://github.com/micro-editor/tcell/tree/legacy) branch contains more or less the same as [^1]

niten94 commented:
Sorry, I was trying to say that I will leave a comment in zyedidia/micro#3593 about the questions brought up in the description. Rebasing with a new version may take a while and it seems like permission has been given to maintain forks and move them to the organization, so I do not see any problem maintaining the current version used there. I have not submitted the pull request to `micro-editor/tcell` yet, so I will create it there.

niten94 commented:
The pull request was closed because I tried renaming the branch. I have not compared the file mentioned at this part in https://github.com/zyedidia/tcell/pull/29#issuecomment-2567023197 a lot yet, but I have classes so I may not be able to open the pull request in `micro-editor` early:

> However, I realized there is this old version of `tscreen_unix.go` similar with this pull request so I could adjust to it if it is better: https://github.com/gdamore/tcell/blob/e9095fe4f117459e3d961ea4eb3f55bf8f45e91c/tscreen_unix.go

JoeKar commented:
Summarizing it (as far as possible) inside a tscreen_unix.go doesn't seem to be the wrong approach, especially since it was also done upstream.
The only important thing is, that the change/PR must be created against the [legacy](https://github.com/micro-editor/tcell/tree/legacy)-branch instead of the master-branch.

niten94 commented:
There doesn't seem to be differences with the original code and functions in `golang.org/x/term`, so I have created a pull request using `x/term` like the old version of `tscreen_unix.go` I mentioned: micro-editor/tcell#2

There will be more changes in code structure if the commit is cherry picked so I replaced the calls instead.
